packages/dds/tree/src/feature-libraries/indexing/anchorTreeIndex.ts

```diff
@@ -0,0 +1,417 @@
+	/**
+	 * The actual index from keys to anchor nodes.
+	 */
+	private readonly nodes = new Map<TKey, AnchorNode[]>();
+	/**
+	 * Maintained for efficient removal of anchor nodes from the index when updating their keys
+	 */
+	private readonly anchorKeys = new Map<AnchorNode, TKey>();
+	/**
+	 * Keeps track of anchors for disposal.
+	 */
+	private readonly anchors = new Map<AnchorNode, Anchor[]>();
+
+	/**
+	 * @param forest - the forest that is being indexed
+	 * @param indexer - a function that retrieves the key finder based on a given schema or undefined if the schema does not have an associated key finder
+	 * @param getValue - a function that returns the value or undefined given at least one anchor node
+	 * @param checkTreeStatus - a function that gets the tree status from an anchor node, used for filtering out detached nodes
+	 */
+	public constructor(
+		private readonly forest: IForestSubscription,
+		private readonly indexer: (
+			schemaId: TreeNodeSchemaIdentifier,
+		) => KeyFinder<TKey> | undefined,
+		private readonly getValue: (anchorNodes: TreeIndexNodes<AnchorNode>) => TValue | undefined,
+		private readonly checkTreeStatus: (node: AnchorNode) => TreeStatus | undefined,
+	) {
+		this.forest.registerAnnouncedVisitor(this.acquireVisitor.bind(this));
+
+		const detachedFieldKeys: FieldKey[] = [];
+		const detachedFieldsCursor = forest.getCursorAboveDetachedFields();
+		forEachField(detachedFieldsCursor, (field) => {
+			detachedFieldKeys.push(field.getFieldKey());
+		});
+
+		// index all existing trees (this includes the primary document tree and all other detached/removed trees)
+		for (const fieldKey of detachedFieldKeys) {
+			const cursor = forest.allocateCursor();
+			forest.tryMoveCursorToField({ fieldKey, parent: undefined }, cursor);
+			this.indexField(cursor);
+			cursor.free();
+		}
+	}
+
+	/**
+	 * Creates an announced visitor that responds to edits to the forest and updates the index accordingly.
+	 */
+	private acquireVisitor(): AnnouncedVisitor {
+		let parentField: FieldKey | undefined;
+		let parent: UpPath | undefined;
+
+		return createAnnouncedVisitor({
+			afterCreate: (content: ProtoNodes, destination: FieldKey) => {
+				const detachedCursor = this.forest.allocateCursor();
+				this.forest.tryMoveCursorToField(
+					{ fieldKey: destination, parent: undefined },
+					detachedCursor,
+				);
+				this.indexField(detachedCursor);
+				detachedCursor.free();
+			},
+			beforeReplace: () => {
+				assert(parent !== undefined, "must have a parent");
+				const cursor = this.forest.allocateCursor();
+				this.forest.moveCursorToPath(parent, cursor);
+				const anchor = cursor.buildAnchor();
+				const anchorNode = this.forest.anchors.locate(anchor);
+				if (anchorNode !== undefined) {
+					const previousKey = this.anchorKeys.get(anchorNode);
+					if (previousKey !== undefined) {
+						this.removeAnchor(anchorNode, previousKey);
```

Comment:
what is this doing? docs might be helpful here :)
I thought we had decided that the removed content didn't need to be removed here? also, if it does, shouldn't it be recursive?